src/transport/CertificateAuthenticatedSessionEstablishment.cpp

```diff
@@ -0,0 +1,415 @@
+    {
+        opaque initiator_identity;
+        opaque signature;
+    };
+
+    struct SigmaR3
+    {
+        SigmaR3Encrypted encrypted_r3;
+    };
+    */
+
+    msg_R3->SetDataLength(data_len);
+
+    // Call delegate to send the Msg3 to peer
+    err = AttachHeaderAndSend(Protocols::SecureChannel::MsgType::CASE_SigmaR3, std::move(msg_R3));
+    SuccessOrExit(err);
+
+    ChipLogDetail(Inet, "Sent SigmaR3 msg");
+
+    mPairingComplete = true;
+
+    // Call delegate to indicate pairing completion
+    mDelegate->OnPairingComplete();
+
+exit:
+
+    if (err != CHIP_NO_ERROR)
+    {
+        SendErrorMsg(SigmaErrorType::kUnexpected);
+    }
+    return err;
+}
+
+CHIP_ERROR CertificateAuthenticatedSessionEstablishment::HandleSigmaR3(const PacketHeader & header,
+                                                                       const System::PacketBufferHandle & msg)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    ChipLogDetail(Inet, "Received SigmaR3 msg");
+
+    // ToDo: Verify received SigmaR3 message.
+
+    mNextExpectedMsg = Protocols::SecureChannel::MsgType::CASE_SigmaErr;
+
+    VerifyOrExit(header.GetSourceNodeId().ValueOr(kUndefinedNodeId) == mConnectionState.GetPeerNodeId(),
+                 err = CHIP_ERROR_WRONG_NODE_ID);
+    VerifyOrExit(header.GetEncryptionKeyID() == mConnectionState.GetPeerKeyID(), err = CHIP_ERROR_INVALID_KEY_ID);
+
+    mPairingComplete = true;
+
+    // Call delegate to indicate pairing completion
+    mDelegate->OnPairingComplete();
+
+exit:
+
+    if (err != CHIP_NO_ERROR)
+    {
+        SendErrorMsg(SigmaErrorType::kUnexpected);
+    }
+    return err;
+}
+
+void CertificateAuthenticatedSessionEstablishment::SendErrorMsg(SigmaErrorType errorCode)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    System::PacketBufferHandle msg;
+    uint16_t msglen      = sizeof(SigmaErrorMsg);
+    SigmaErrorMsg * pMsg = nullptr;
+
+    msg = System::PacketBuffer::NewWithAvailableSize(msglen);
+    VerifyOrExit(!msg.IsNull(), err = CHIP_SYSTEM_ERROR_NO_MEMORY);
+
+    pMsg        = reinterpret_cast<SigmaErrorMsg *>(msg->Start());
+    pMsg->error = errorCode;
+
+    msg->SetDataLength(msglen);
+
+    err = AttachHeaderAndSend(Protocols::SecureChannel::MsgType::CASE_SigmaErr, std::move(msg));
+    SuccessOrExit(err);
+
+exit:
+    Clear();
+}
+
+void CertificateAuthenticatedSessionEstablishment::HandleErrorMsg(const PacketHeader & header,
+                                                                  const System::PacketBufferHandle & msg)
+{
+    // Error message processing
+    const uint8_t * buf  = msg->Start();
+    size_t buflen        = msg->DataLength();
+    SigmaErrorMsg * pMsg = nullptr;
+
+    VerifyOrExit(buf != nullptr, ChipLogError(Inet, "Null error msg received during pairing"));
+    VerifyOrExit(buflen == sizeof(SigmaErrorMsg), ChipLogError(Inet, "Error msg with incorrect length received during pairing"));
+
+    pMsg = reinterpret_cast<SigmaErrorMsg *>(msg->Start());
```

Comment:
This relies on both sender and receiver being same-endian or `SigmaErrorMsg` being one octet.  We should either `static_assert` the latter or do endian-aware reading/writing.